_source/_posts/2024-02-29-net-scim.md

```diff
@@ -0,0 +1,1898 @@
+---
+
+Several challenges exist in provisioning users for a growing company across systems and applications. It can be time-consuming, error-prone, challenging to sync users across domains, and potentially a security risk to keep track of existing profiles. Fortunately, a protocol called SCIM (System for Cross-domain Identity Management) standardizes user information across systems, making it possible to sync users. Also, combined with a SCIM-compliant Identity Provider (IdP), it securely automates common user lifecycle operations.
+In this tutorial, I'll walk you through building a .NET SCIM server with the CRUD (Create, Read, Update, Delete) Restful API endpoints outlined in the SCIM protocol. We'll then test the server using Swagger User Interface before integrating it with Okta, a SCIM compliant IdP. 
+
+## Use SCIM to manage users
+[SCIM](https://datatracker.ietf.org/doc/html/rfc7644) is an HTTP-based protocol that defines a set of standard endpoints and schemas that developers can use to overcome complexity in implementing user lifecycle management operations across different systems.
+
+## Components of a SCIM integration
+There are two primary components for a SCIM integration. The first is the *SCIM server*, which is a server that has implemented the endpoints as described in the SCIM spec - for example, User endpoints, Group endpoints, Schemas endpoints, etc. The applications team typically develops these endpoints, which leaves the IdP to manage the user profiles. This blog will detail how to create a sample SCIM server using .NET. The other component is the *SCIM client*, which makes SCIM-compliant HTTP calls to a SCIM server to exchange user profile information. Okta (the IdP), in our use case, will act as a SCIM client and call the .NET SCIM server we'll build using this tutorial.
+
+## Build a .NET SCIM server
+
+**Prerequisites**
+- dotnet SDK (I used [dotnet 8](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) in this demo)
+- Code Editor (I used Visual Studio Code) 
+- [Okta CLI](https://cli.okta.com/)
+
+{% include toc.md %}
+
+### Create a .NET project
+Create a directory called `okta-scim-server-dotnet`. Navigate to the directory in your terminal, then run the following commands:
+- `dotnet new webapi` (Creates a new API project)
+- `dotnet dev-certs https --trust` (Trust the self-signed TLS certs)
+
+### Test the .NET project using Swagger UI
+- Run the project using the command `dotnet watch --launch-profile https`
+- At this point using the *https://localhost:7094/swagger/index.html* you will be able to see the Swagger UI. Typically, a browser tab will automatically open; if not, manually check for the URL to the Swagger UI under `Properties/launchSettings.json` and copy/paste this URL onto the browser. 
+
+### Setup Okta (the IdP)
+{% include setup/cli.md type="spa" framework="ASP.NET Core" loginRedirectUri="http://localhost:7094/swagger/oauth2-redirect.html" logoutRedirectUri="http://localhost:7094" %}
```

Comment:
```suggestion
{% include setup/cli.md type="spa" loginRedirectUri="http://localhost:7094/swagger/oauth2-redirect.html" logoutRedirectUri="http://localhost:7094" %}
```
Removing because it's not needed for rendering and it just leaves confusion since ASP .NET isn't a SPA